Report the actual Cassandra Version

wallrj opened the pull request:
* Add a Pilot.Status.Cassandra.Version field.
* Add a nodetool.Version method which queries Cassandra version via JMX / Jolokia.
* A CassandraVersion type that wraps semver.Version in order to add a special case for versions without a patch number.
* Update the Pilot.Status.Cassandra.Version field during calls to Pilot.Sync.
* An E2E test for version reporting.

Fixes: #262 

**Release note**:
```release-note
NONE
```

wallrj commented:
Rebased after closing #244

kragniz commented on pkg/apis/navigator/v1alpha1/types.go:
```diff
@@ -299,6 +301,11 @@ type ElasticsearchPilotStatus struct {
 }
 
+type CassandraPilotStatus struct {
+	// Version as reported by the Elasticsearch process
```
Elasticsearch -> Cassandra

kragniz commented on pkg/pilot/cassandra/v3/pilot.go:
```diff
@@ -80,9 +80,24 @@ func (p *Pilot) CmdFunc(pilot *v1alpha1.Pilot) (*exec.Cmd, error) {
+		return err
+	}
+	pilot.Status.Cassandra.Version = version
+	_, err = p.navigatorClient.Navigator().Pilots(pilot.Namespace).UpdateStatus(pilot)
```
Should we only perform the update if the version differs?

wallrj commented:
* Only update if version changed
* And if nodetool.Version fails, attempt to set the .Status.Version to nil
* Added some unit tests for Pilot.Sync

munnerz commented on hack/e2e.sh:
```diff
@@ -230,6 +230,16 @@ function test_cassandracluster() {
+         stdout_equals "${CASS_VERSION}" \
+         kubectl --namespace "${namespace}" \
+         get pilots \
+         --output 'jsonpath={.items[*].status.cassandra.version}'
```
'*' here will select versions on all pilots, no? How does this work if there are 2 pilots returned? (or perhaps there's not 2 pilots at the point this test is run?)

wallrj replied:
There's only one pilot at this point, but you're right it'd be better give index-0

munnerz commented on pkg/apis/navigator/types.go:
```diff
@@ -206,6 +208,12 @@ type ElasticsearchPilotStatus struct {
+type CassandraPilotStatus struct {
+	// The version reported by the Cassandra process.
+	// nil indicates that the version is not yet known / not yet reported.
+	Version *semver.Version
```
This comment is best on the v1alpha1 type (especially the extra info about nil). We won't be generating docs from internal types, only the external versions.

munnerz commented on pkg/cassandra/version/version.go:
```diff
@@ -0,0 +1,45 @@
+		err2 := v.Version.Set(s + ".0")
+		if err2 != nil {
+			return err1
+		}
```
This is a very confusing dance of `err` vars. Can we flatten this to use `err` as much as possible, and if it isn't possible, give the second error a more helpful name than `err1`?

munnerz commented on pkg/cassandra/version/version.go:
```diff
@@ -0,0 +1,45 @@
+	}
+	err1 := v.Version.Set(s)
+	if err1 != nil {
+		err2 := v.Version.Set(s + ".0")
```
I think we need a more thorough check of the existing version string before assuming we can just add `.0` onto the end. This also wouldn't handle something like `v3.0-alpha.1` correctly (as it'd be parsed as `v3.0-alpha.1.0`.

munnerz commented on pkg/cassandra/version/version.go:
```diff
@@ -0,0 +1,45 @@
+	return nil
+}
+
+var _ json.Unmarshaler = &Version{}
```
We also need to implement a marshaller that ensures that unmarshalling and then remarshalling is idempotent imo. This could cause confusion for users otherwise.

OTOH, perhaps we should actually disallow this invalid semver within Cassandra and *always* require valid semvers on our resources. If Cassandra reports an invalid semver, we could 'repair' it within the Pilot itself. Not sure of the implications of this, but given Cassandra is semver-like, it might work well & also saves our idempotentcy (I invented a new word) issues.

wallrj replied:
I've switched to using Cassandra version throughout.
And added the MarshalJSON method.

munnerz commented on pkg/cassandra/nodetool/fake/fake.go:
```diff
@@ -0,0 +1,40 @@
+import (
+	"fmt"
+
+	"github.com/coreos/go-semver/semver"
```
Should we be using our internal cassandra version type here?

munnerz commented on pkg/cassandra/nodetool/nodetool_test.go:
```diff
@@ -356,3 +358,26 @@ func TestNodeToolStatus(t *testing.T) {
+	if ntErr != nil {
+		t.Error(ntErr)
+	}
+	t.Log(ntResponse)
```
We should probably verify this version is as expected, and not just that the function doesn't error. Also should we have a test for a case where the `Version` function *does* error?

munnerz commented on pkg/pilot/cassandra/v3/pilot.go:
```diff
@@ -80,9 +82,39 @@ func (p *Pilot) CmdFunc(pilot *v1alpha1.Pilot) (*exec.Cmd, error) {
+	pilot, err := p.pilotLister.Pilots(pilot.Namespace).Get(pilot.Name)
+	if err != nil {
+		return err
+	}
```
Don't need to obtain a copy of `pilot` - you already have one as the arg to the sync func which was just retrieved from the lister (in the callsite for this function)

munnerz commented on pkg/pilot/cassandra/v3/pilot.go:
```diff
@@ -80,9 +82,39 @@ func (p *Pilot) CmdFunc(pilot *v1alpha1.Pilot) (*exec.Cmd, error) {
+		errs = append(errs, err)
+		pilotCopy.Status.Cassandra.Version = nil
+	}
+	if !reflect.DeepEqual(pilot, pilotCopy) {
```
We should avoid DeepEqual as much as possible - it's slow and expensive. A simple bool for needsUpdate would suffice, especially as this block is quite simple and small at this point.

munnerz commented on pkg/pilot/cassandra/v3/pilot_test.go:
```diff
@@ -0,0 +1,147 @@
+			assertions: func(t *testing.T, state *controllers.State) {
+				CheckPilotVersion(
+					localPilot.Name, localPilot.Namespace, nil, t, state,
+				)
```
I like this `assertions` field 😄

wallrj commented:
Thanks for the review @munnerz 

I've updated the patch to use the Cassandra `version.Version` through out....including in the Navigator API.
If a user specifies `CassandraCluster.Spec.Version: 3.9`  , 
 * the default image will be 3.9.0.
 * Pilot.Status.Cassandra.version will be 3.9

I'll add comparison methods in a followup branch, when I implement the Cassandra Upgrade action.

I think I've addressed all your other comments.

Please take another look it the E2E tests pass.

wallrj commented:
I think those test failures were caused by me using version: 3.11 in the E2E tests....which got pulled as 3.11.0 and failed without any debug logs.

I've added a `trap dump_debug_logs _artifacts EXIT` to ensure the logs get dumped, even if the script exits because of a `set -e` error....or indeed when the script exits without a failure.

Which should help debug similar test failures in future.

wallrj commented:
While working on the Upgrade action branch I kept seeing `pilot.status.cassandra` and `pilot.status.cassandra.version` flickering....sometimes it'd be there and next second it'd be nil.

I think the problem was that I was performing my own pilot status update...not realising that the generic pilot also does a status update.

kragniz commented on pkg/cassandra/version/version.go:
```diff
@@ -0,0 +1,110 @@
+// In Navigator we parse the Cassandra version string as early as possible, into a similar Cassandra Version object.
+// This also fixes the missing Patch number and stores the version internally as a semver.
+// It also keeps a reference to the original version string so that we can report that in our API.
+// So that the version reported in our API matches the version that an administrator expects.
```
+1 good comments for context
